### PR TITLE
PLAT-1188 Implement cascading workflow auto transitions

### PR DESCRIPTION
**Pull request by alexander-schmidt**
- Implemented `WorkflowAutoTransitionCascadedExecutor[Test]`.
- Intentionally not yet used in `WorkflowAutoTransitionExecutionProgram`. This will be changed in a separate PR. We should probably even introduce a flag for that.

### CODE REVIEW

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/AGWorkflowItem.java [hunk 1]
```diff
@@ -51,6 +52,15 @@ public Collection<Set<AGWorkflowTransitionExecution>> loadTransitionExecutionSet
 		return loadTransitionExecutionMap().getSets();
 	}
 
+	/**
+	 * Executes the longest possible cascade of auto transitions for this
+	 * {@link AGWorkflowItem}.
+	 */
+	public void executeCascadingAutoTransitions() {
```
Comment (oliver-richers) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/AGWorkflowItem.java:59]: - the name may erroneously imply for someone that only cascading auto-transitions will be executed
- suggested names
  - `executeAllAutoTransitions`
  - `executeAllPossibleAutoTransitions`
  - `executeAutoTransitionsDeep`
  - `executeAutoTransitionsRecursively`

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/IWorkflowableObject.java [hunk 2]
```diff
@@ -34,6 +34,22 @@ default boolean isWorkflowStarted() {
...
+	/**
+	 * Executes the longest possible cascade of auto transitions for the
+	 * {@link AGWorkflowItem}.
+	 * <p>
+	 * If this {@link IWorkflowableObject} is not associated with a
+	 * {@link AGWorkflowItem}, nothing will happen.
+	 *
+	 * @see #getWorkflowItem()
+	 */
+	default void executeCascadingAutoTransitions() {
```
Comment (oliver-richers) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/IWorkflowableObject.java:46]: same though concerning the name as in `AGWorkflowItem`

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/execution/auto/WorkflowAutoTransitionCascadedExecutor.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		this.executor = new WorkflowAutoTransitionExecutor(item);
+	}
+
+	/**
+	 * Defines the maximum length of an auto transition cascade.
+	 *
+	 * @param cascadeLengthLimit
+	 * @return this
+	 */
+	public WorkflowAutoTransitionCascadedExecutor setCascadeLengthLimit(int cascadeLengthLimit) {
```
Comment (oliver-richers) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/execution/auto/WorkflowAutoTransitionCascadedExecutor.java:40]: possible alternative naming `setRecursionLimit`

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/execution/auto/WorkflowAutoTransitionCascadedExecutor.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	 * When the length limit is reached while processing a cascade, an
+	 * {@link Exception} is thrown.
+	 */
+	public void evaluateAndExecuteCascaded() {
+
+		int counter = 0;
+		while (evaluateAndExecute()) {
+			counter++;
+			if (counter >= cascadeLengthLimit) {
+				throw new SofticarException("Exceeded the maximum cascade length of %s.", cascadeLengthLimit);
```
Comment (oliver-richers) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/execution/auto/WorkflowAutoTransitionCascadedExecutor.java:60]:
```suggestion
				throw new SofticarException("Exceeded the maximum auto transition cascade length of %s.", cascadeLengthLimit);
```

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/execution/auto/WorkflowAutoTransitionExecutor.java [hunk 5]
```diff
@@ -10,34 +10,55 @@
...
+	/**
+	 * Executes an automatic {@link AGWorkflowTransition} for the
+	 * {@link AGWorkflowItem}.
+	 * <p>
+	 * Throws {@link SofticarUserException} if there is more than one
+	 * {@link AGWorkflowTransition} with a valid precondition.
+	 * <p>
+	 * Returns <i>true</i> if there is only one {@link AGWorkflowTransition}
+	 * with a valid precondition, and if that transition was successfully
+	 * executed.
```
Comment (oliver-richers) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/execution/auto/WorkflowAutoTransitionExecutor.java:36]:
```suggestion
	 * with a valid precondition and a source node equal to the current node
	 * of the item, and if that transition was successfully executed.
```

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/AGWorkflowItem.java (hunk 1 above)
Comment (alexander-schmidt) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/AGWorkflowItem.java:59]: Oped for `executeAllAutoTransitions`.

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/IWorkflowableObject.java (hunk 2 above)
Comment (alexander-schmidt) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/IWorkflowableObject.java:46]: see above

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/execution/auto/WorkflowAutoTransitionCascadedExecutor.java (hunk 3 above)
Comment (alexander-schmidt) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/execution/auto/WorkflowAutoTransitionCascadedExecutor.java:40]: Everything so far is dubbed "cascade". I think the the introduction of a synonym would not be beneficial.

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/execution/auto/WorkflowAutoTransitionCascadedExecutor.java (hunk 4 above)
Comment (alexander-schmidt) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/execution/auto/WorkflowAutoTransitionCascadedExecutor.java:60]: ok

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/execution/auto/WorkflowAutoTransitionExecutor.java (hunk 5 above)
Comment (alexander-schmidt) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/execution/auto/WorkflowAutoTransitionExecutor.java:36]: ok

**Review by oliver-richers — APPROVED** [review]

**Pull request merged**